metagraph/src/graph/annotated_graph_algorithm.cpp

```diff
@@ -15,249 +12,416 @@ namespace graph {
+    size_t width = sdsl::bits::hi(num_labels) + 1;
+    sdsl::bit_vector indicator(anno_graph.get_graph().max_index() + 1, false);
 
-    std::mutex backup_mutex;
-    constexpr std::memory_order memorder = std::memory_order_relaxed;
+    // the in and out counts are stored interleaved
+    sdsl::int_vector<> counts = aligned_int_vector(indicator.size() * 2, 0, width, 16);
 
-    std::atomic_thread_fence(std::memory_order_release);
-    #pragma omp parallel for num_threads(num_threads) schedule(dynamic)
-    for (size_t j = 0; j < labels_in.size(); ++j) {
-        const std::string &label_in = labels_in[j];
-        anno_graph.call_annotated_nodes(label_in, [&](DeBruijnGraph::node_index i) {
-            atomic_fetch_and_add(counts, i, 1, backup_mutex, memorder);
-        });
-    }
-    std::atomic_thread_fence(std::memory_order_acquire);
+    const auto &label_encoder = anno_graph.get_annotation().get_label_encoder();
+    const auto &binmat = anno_graph.get_annotation().get_matrix();
 
-    // correct the width of counts, making it single-width
-    counts.width(width);
+    std::vector<uint64_t> label_in_codes;
+    label_in_codes.reserve(labels_in.size());
+    std::vector<uint64_t> label_out_codes;
+    label_out_codes.reserve(labels_out.size());
+    for (const std::string &label_in : labels_in) {
+        label_in_codes.push_back(label_encoder.encode(label_in));
+    }
+    for (const std::string &label_out : labels_out) {
+        label_out_codes.push_back(label_encoder.encode(label_out));
+    }
 
+    std::mutex vector_backup_mutex;
     std::atomic_thread_fence(std::memory_order_release);
-    #pragma omp parallel for num_threads(num_threads) schedule(dynamic)
-    for (size_t j = 0; j < labels_out.size(); ++j) {
-        const std::string &label_out = labels_out[j];
-        anno_graph.call_annotated_nodes(label_out, [&](DeBruijnGraph::node_index i) {
-            atomic_fetch_and_add(counts, (i << 1) + 1, 1, backup_mutex, memorder);
-        });
-    }
-    std::atomic_thread_fence(std::memory_order_acquire);
+    bool parallel = num_threads > 1;
+    binmat.call_columns(label_in_codes,
+        [&](auto, const bitmap &rows) {
+            rows.call_ones([&](auto r) {
+                node_index i = AnnotatedDBG::anno_to_graph_index(r);
+                set_bit(indicator.data(), i, parallel, MO_RELAXED);
+                atomic_fetch_and_add(counts, i * 2, 1, vector_backup_mutex, MO_RELAXED);
+            });
+        },
+        num_threads
+    );
 
-    // set the width to be double again
-    counts.width(width << 1);
+    binmat.call_columns(label_out_codes,
+        [&](auto, const bitmap &rows) {
+            rows.call_ones([&](auto r) {
+                node_index i = AnnotatedDBG::anno_to_graph_index(r);
+                set_bit(indicator.data(), i, parallel, MO_RELAXED);
+                atomic_fetch_and_add(counts, i * 2 + 1, 1, vector_backup_mutex, MO_RELAXED);
+            });
+        },
+        num_threads
+    );
```

Comment:
Would be good to merge these two to reduce everything to a single call of call_columns. For hierarchically compressed representation this can greatly reduce the run time.

1. build a total list of columns: std::vector<Column> and indicators sdsl::vector<int> -- in/out/both
2. query all columns and update the counts according to the indicators (in/out/both)

Reply:
good point, I implemented it this way. Since in and out are disjoint, I just did this using a std::vector<bool>